runner/utils/driver/istio/secrets.go

```diff
@@ -0,0 +1,162 @@
+			RootPath: v.ObjectMeta.Annotations["rootPath"],
+			Data:     make(map[string]string, len(v.Data)),
+		}
+		for k1 := range v.Data {
+			s.Data[k1] = ""
+		}
+		listOfSecrets[i] = s
+	}
+	return listOfSecrets, nil
+}
+
+// DeleteSecret is used to delete secrets!
+func (i *Istio) DeleteSecret(projectID string, secretName string) error {
+	err := i.kube.CoreV1().Secrets(projectID).Delete(secretName, &metav1.DeleteOptions{})
+	if err != nil {
+		logrus.Errorf("secret could not be deleted- %s", err)
+		return fmt.Errorf("secret could not be deleted - %s", err)
+	}
+	return nil
+}
+
+// SetKey adds a new secret key-value pair
+func (i *Istio) SetKey(projectID string, secretValObj *model.SecretValue, secretName string, secretKey string) error {
+	if secretName == "" || secretValObj.Value == "" {
+		logrus.Errorf("Empty key/value provided. Key not set")
+		return fmt.Errorf("key/value not provided; got (%s,%s)", secretName, secretValObj.Value)
+	}
+	// encoding secret value to base64
+	encSecret := b64.StdEncoding.EncodeToString([]byte(secretValObj.Value))
+	//Get secret and then check type
+	kubeSecret, err := i.kube.CoreV1().Secrets(projectID).Get(secretName, metav1.GetOptions{})
+
+	if kubeErrors.IsNotFound(err) {
+		return fmt.Errorf("secret with name (%s) does not exist- %s", secretName, err)
+	} else if err == nil {
+		//Add secret key-value
+		switch kubeSecret.Type {
+		case v1.SecretTypeDockerConfigJson:
+			return fmt.Errorf("setKey operation cannot be performed on secrets with type -docker")
+		case v1.SecretTypeOpaque:
+			kubeSecret.Data[secretKey] = []byte(encSecret)
+		default:
+			//Throw error
+			return fmt.Errorf("invalid secret type - %s", err)
+		}
+
+		//Update the secret
+		_, err := i.kube.CoreV1().Secrets(projectID).Update(kubeSecret)
+		if err != nil {
+			return fmt.Errorf("secret-key could not be set - %s", err)
+		}
+	}
+	return nil
+}
+
+// DeleteKey is used to delete a key from the secret!
+func (i *Istio) DeleteKey(projectID string, secretName string, secretKey string) error {
+	// Get secret
+	kubeSecret, err := i.kube.CoreV1().Secrets(projectID).Get(secretName, metav1.GetOptions{})
+
+	if kubeErrors.IsNotFound(err) {
+		return fmt.Errorf("secret with name (%s) does not exist- %s", secretName, err)
+	} else if err == nil {
+		// check the type of secret (docker/opaque)
+		switch kubeSecret.Type {
+		case v1.SecretTypeDockerConfigJson:
+			return fmt.Errorf("deleteKey operation cannot be performed on secrets with type -docker")
+		case v1.SecretTypeOpaque:
+			delete(kubeSecret.Data, secretKey)
+		default:
+			// Throw error
+			return fmt.Errorf("invalid secret type - %s", err)
+		}
+		// Update the secret
+		_, err := i.kube.CoreV1().Secrets(projectID).Update(kubeSecret)
+		if err != nil {
+			return fmt.Errorf("secret-key could not be deleted - %s", err)
+		}
+	}
+	return nil
+}
+
+// helper function
+func generateSecret(projectID string, secret *model.Secret) *v1.Secret {
```

Comment:
For docker secrets we were supposed to create a json body and base64 encode that remember?